File: document/src/vespa/document/base/documentcalculator.cpp
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "documentcalculator.h"
#include <vespa/document/bucket/bucketidfactory.h>
#include <vespa/document/select/compare.h>
#include <vespa/document/select/parser.h>
#include <vespa/document/select/valuenode.h>
#include <vespa/document/select/variablemap.h>
#include <vespa/vespalib/util/exceptions.h>

namespace document {

DocumentCalculator::DocumentCalculator(
        const DocumentTypeRepo& repo,
        const vespalib::string & expression)
{
    BucketIdFactory factory;
    select::Parser parser(repo, factory);
    _selectionNode = parser.parse(expression + " == 0");
}

DocumentCalculator::~DocumentCalculator() { }

double
DocumentCalculator::evaluate(const Document& doc, std::unique_ptr<select::VariableMap> variables)
{
    select::Compare& compare(static_cast<select::Compare&>(*_selectionNode));
    const select::ValueNode& left = compare.getLeft();

    select::Context context(doc);
    context.setVariableMap(std::move(variables));
    std::unique_ptr<select::Value> value = left.getValue(context);

    select::NumberValue* num = dynamic_cast<select::NumberValue*>(value.get());

    if (!num) {
        throw vespalib::IllegalArgumentException("Expression could not be evaluated - some components of the expression may be missing");
    }

    return num->getCommonValue();
}

}

File: document/src/main/java/com/yahoo/document/DocumentCalculator.java
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.document;

import com.yahoo.document.select.Context;
import com.yahoo.document.select.parser.ParseException;
import com.yahoo.document.select.parser.SelectInput;
import com.yahoo.document.select.parser.SelectParser;
import com.yahoo.document.select.rule.ComparisonNode;

import java.util.Map;

/**
 * @author Thomas Gundersen
 */
public class DocumentCalculator {

    private ComparisonNode comparison;

    public DocumentCalculator(String expression) throws ParseException {
        SelectParser parser = new SelectParser(new SelectInput(expression + " == 0"));
        comparison = (ComparisonNode)parser.expression();
    }

    public Number evaluate(Document doc, Map<String, Object> variables) {
        Context context = new Context(new DocumentPut(doc));
        context.setVariables(variables);

        try {
            Object o = comparison.getLHS().evaluate(context);

            if (Double.isInfinite(((Number)o).doubleValue())) {
                throw new IllegalArgumentException("Expression evaluated to an infinite number");
            }
            return ((Number)o).doubleValue();
        } catch (ArithmeticException e) {
            throw new IllegalArgumentException("Arithmetic exception " + e.getMessage(), e);
        }
    }

}
